### PR TITLE
persist: hardcode crossbeam_channel for "listeners"

### PR DESCRIPTION
**Pull request by mkysel**
Refactor listeners to use a crossbeam_channel::sender instead of an arbitrary lambda.

### Motivation

Executing any code in the persist hot path is dangerous and should not be possible.

### CODE REVIEW

File: src/persist/src/operators/stream.rs [hunk 1]
```diff
@@ -876,6 +874,7 @@ mod tests {
     }
 
     #[test]
+    #[ignore]
```
Comment (mkysel) [inline src/persist/src/operators/stream.rs:877]: @danhhz you said that you have opinions on how this should be done.

Comment (aljoscha) [reply src/persist/src/operators/stream.rs:877]: I think it's hard (or impossible) to do it "from the outside". If we push all the `Seal` to one channel we loose the information about where they are from, if we use two channels, we lose the ordering information.

Internally, we have `SegNo`s that tell us the order in which we apply operations. We can verify that we get the expected ordering of seal calls, for the to streams, but I don't know how easy it is to get at these from the outside.

Comment (danhhz) [reply src/persist/src/operators/stream.rs:877]: we could do it with 3 channels and an async background task. each listen would get a channel, then the task would repeatedly read from whichever channel has something available and write it to the third channel, which would then contain exactly what the single channel here does now. it's a lot of complexity, but this test is worth it if we can't think of anything better

Comment (mkysel) [reply src/persist/src/operators/stream.rs:877]: is there a guarantee that the background task will read the two independent channels in the same order?

Comment (danhhz) [reply src/persist/src/operators/stream.rs:877]: no but that's not a requirement IMO (I'm curious if @aljoscha agrees). the important thing is that any order that we do observe upholds the invariants that are checked later in the test https://github.com/MaterializeInc/materialize/pull/9107/files#diff-484921e74ba303073fe90a8e8442998e6df16e12c8734710cd1ed3c2b4f2df3bR962-R976

Comment (aljoscha) [reply src/persist/src/operators/stream.rs:877]: The order in which Seals from the two channels arrive would matter. Say both channels contain a `Seal(1)`, it's important to know which one was put in the channel first. And I think that information is lost.

Comment (aljoscha) [reply src/persist/src/operators/stream.rs:877]: (Sorry if this is going very slowly, we can chat about it when we're both online at the same time)

**Comment by mkysel** [comment]
I will return to this review on monday

**Comment by mkysel** [comment]
I have rewritten the test to use the suggested background aggregator. Removed the `draft` state

File: src/persist/src/indexed/mod.rs [hunk 2]
```diff
@@ -834,21 +833,22 @@ impl<L: Log, B: Blob> Indexed<L, B> {
         }
 
         for (id, updates) in updates {
-            if let Some(listen_fns) = self.listeners.get(&id) {
-                if listen_fns.len() == 1 {
-                    listen_fns[0].0(ListenEvent::Records(updates));
+            if let Some(listeners) = self.listeners.get(&id) {
+                // writer to the listeners, this will plumb into the persist code
```
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:837]: not sure this comment is getting us much, let's remove it

File: src/persist/src/operators/stream.rs [hunk 3]
```diff
@@ -893,31 +892,33 @@ mod tests {
...
 
-        {
-            let listen_tx = listen_tx.clone();
-            let listen_fn = ListenFn(Box::new(move |e| {
-                match e {
-                    ListenEvent::Sealed(t) => listen_tx.send(Sealed::Primary(t)).unwrap(),
+        // best effort loop to categorize and forward incoming messages to an aggregate channel
+        // for later test validation logic
+        let (listen_aggregate_tx, listen_aggregate_rx) = crossbeam_channel::unbounded();
+        let _ = thread::spawn(move || loop {
```
Comment (danhhz) [inline src/persist/src/operators/stream.rs:904]: let's use an async runtime for this, so we can use something like https://docs.rs/futures/0.3.17/futures/macro.select.html to repeatedly grab from whichever is available first without the hot loop. we should also have that bit wait until both channels are closed before returning and block the test on this async task finished to ensure everything is cleaned up at the end

**Pull request closed without merging**